### PR TITLE
Add threshold for FS Calls and print on actions taking over those thresholds

### PR DESCRIPTION
**Pull request by liranmauda**
### Explain the changes

- Add threshold for FS Calls and print on actions taking over those thresholds

Signed-off-by: liranmauda <liran.mauda@gmail.com>

### CODE REVIEW

File: src/native/fs/fs_napi.cpp [hunk 1]
```diff
@@ -77,6 +77,7 @@ struct FSWorker : public Napi::AsyncWorker
     std::string _backend;
     std::string _desc;
     int _errno;
+    int thresholdInSec = 1 * 60;
```
Comment (liranmauda) [inline src/native/fs/fs_napi.cpp:80]: Probably 1 min is way too much.
@guymguym @jeniawhite What do you think the threshold should be?

Comment (jeniawhite) [reply src/native/fs/fs_napi.cpp:80]: I would expect to measure milliseconds and not seconds/minutes.
The only thing is that we are measuring the whole work function (includes writing buffers, reading MD, reading directories which include multiple POSIX operations that might take a bit more time).
But still, I think the granularity should be in milliseconds.

File: src/native/fs/fs_napi.cpp [hunk 2]
```diff
@@ -102,7 +103,12 @@ struct FSWorker : public Napi::AsyncWorker
         DBG1("FS::FSWorker::Execute: " << _desc << DVAL(_uid) << DVAL(_gid) << DVAL(_backend));
         ThreadScope tx;
         tx.set_user(_uid, _gid);
+        std::time_t timeBefore = std::time(nullptr);
```
Comment (jeniawhite) [inline src/native/fs/fs_napi.cpp:106]: time_t holds the number of seconds since the epoch.
I'm not sure if that is the granularity that we want to use, I would see if we could have something at least in the millisecond range.

File: src/native/fs/fs_napi.cpp (hunk 1 above)
Comment (liranmauda) [reply src/native/fs/fs_napi.cpp:80]: That's the point, syscall will take milliseconds, but the whole write?
In any case, as this is only a printing, We will set it to 100 milliseconds and if we need (get too many printings) we tweak the threshold.

Comment (liranmauda) [reply src/native/fs/fs_napi.cpp:80]: Testing upload of 25MB file:
```
2021-06-21 07:33:08.302061 [PID-9/TID-9] [L0] FS::FSWorker::Begin: FileWritev _wrap->_path=/nsfs/ns/.noobaa-nsfs_60cf4e763f991c002842dddb/uploads/f20c0854-7a49-43d2-8c31-0f2231cee936 total_len=8388608 buffers_len=512
2021-06-21 07:33:08.302230 [PID-9/TID-18] [L0] timeBefore 1624260788302
2021-06-21 07:33:08.305321 [PID-9/TID-18] [L0] timeAfter 1624260788305
2021-06-21 07:33:08.907173 [PID-9/TID-9] [L0] FS::FSWorker::Begin: FileWritev _wrap->_path=/nsfs/ns/.noobaa-nsfs_60cf4e763f991c002842dddb/uploads/f20c0854-7a49-43d2-8c31-0f2231cee936 total_len=8388608 buffers_len=510
2021-06-21 07:33:08.907354 [PID-9/TID-17] [L0] timeBefore 1624260788907
2021-06-21 07:33:09.399878 [PID-9/TID-17] [L0] timeAfter 1624260789399
2021-06-21 07:33:09.399975 [PID-9/TID-17] [L0] FS::FSWorker::Execute: FileWritev _wrap->_path=/nsfs/ns/.noobaa-nsfs_60cf4e763f991c002842dddb/uploads/f20c0854-7a49-43d2-8c31-0f2231cee936 total_len=8388608 buffers_len=510  took: 492 milliseconds
2021-06-21 07:33:09.799042 [PID-9/TID-9] [L0] FS::FSWorker::Begin: FileWritev _wrap->_path=/nsfs/ns/.noobaa-nsfs_60cf4e763f991c002842dddb/uploads/f20c0854-7a49-43d2-8c31-0f2231cee936 total_len=820261 buffers_len=51
2021-06-21 07:33:09.799737 [PID-9/TID-16] [L0] timeBefore 1624260789799
2021-06-21 07:33:09.899385 [PID-9/TID-16] [L0] timeAfter 1624260789899
```

As `FileWritev` can take 492 milliseconds we will set the threshold to 100 milliseconds

**Review by jeniawhite — APPROVED** [review]

File: src/native/fs/fs_napi.cpp [hunk 3]
```diff
@@ -99,10 +100,22 @@ struct FSWorker : public Napi::AsyncWorker
...
+
         DBG1("FS::FSWorker::Execute: " << _desc << DVAL(_uid) << DVAL(_gid) << DVAL(_backend));
+        if (_desc.find("FileWritev") != std::string::npos ) {
+            thresholdInMilliSec = 100;
+        }
         ThreadScope tx;
         tx.set_user(_uid, _gid);
+        std::time_t timeBefore = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
         Work();
+        std::time_t timeAfter = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:115]: This is not the way to measure elapsed time in cpp.
See https://www.geeksforgeeks.org/measure-execution-time-function-cpp/
TLDR:
```c++
auto start_time = std::chrono::high_resolution_clock::now();
Work();
auto end_time = std::chrono::high_resolution_clock::now();
double took_time = std::chrono::duration<double, std::milli>(end_time - start_time).count();
```

File: src/native/fs/fs_napi.cpp [hunk 4]
```diff
@@ -99,10 +100,22 @@ struct FSWorker : public Napi::AsyncWorker
...
+        if (timeAfter - timeBefore > thresholdInMilliSec) {
+            DBG0("FS::FSWorker::Execute: " << _desc << " took: " << timeAfter - timeBefore << " milliseconds");
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:117]: I expected a `WARNING TOOK TOO LONG` in the text, this looks too informative.

File: src/native/fs/fs_napi.cpp [hunk 5]
```diff
@@ -99,10 +100,22 @@ struct FSWorker : public Napi::AsyncWorker
...
     void Execute() override
     {
+        using std::chrono::duration_cast; 
+        using std::chrono::milliseconds;
+        using std::chrono::system_clock;
+
         DBG1("FS::FSWorker::Execute: " << _desc << DVAL(_uid) << DVAL(_gid) << DVAL(_backend));
+        if (_desc.find("FileWritev") != std::string::npos ) {
+            thresholdInMilliSec = 100;
+        }
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:110]: This is not the way to change behavior of inherited classes.
Instead, use a method called `set_work_threashold(...)` and call it in the ctor of FileWritev.

File: src/native/fs/fs_napi.cpp [hunk 6]
```diff
@@ -77,6 +77,7 @@ struct FSWorker : public Napi::AsyncWorker
     std::string _backend;
     std::string _desc;
     int _errno;
+    int thresholdInMilliSec = 2;
```
Comment (guymguym) [inline src/native/fs/fs_napi.cpp:80]: confusing naming standards - I expected to see:
```
    int _warn_threshold_ms = 2;
```

**Review by guymguym — COMMENTED** [review]
@liranmauda See some comments after already merged. Thanks
